packages/builder/src/components/backend/Datasources/relationshipErrors.js

```diff
@@ -0,0 +1,89 @@
+import { RelationshipTypes } from "constants/backend"
+
+const typeMismatch = "Column type of the foreign key must match the primary key"
+const columnCantExist = "Column name cannot be an existing column"
+const mustBeDifferentTables = "From/to/through tables must be different"
+const primaryKeyNotSet = "Please pick the primary key"
+const throughNotNullable =
+  "Ensure non-key columns are nullable or auto-generated"
+const noRelationshipType = "Please specify a relationship type"
+const tableNotSet = "Please specify a table"
+const foreignKeyNotSet = "Please pick a foreign key"
+const relationshipAlreadyExists =
+  "A relationship between these tables already exists"
+
+function isColumnNameBeingUsed(table, columnName, originalName) {
+  if (!table || !columnName || columnName === originalName) {
+    return false
+  }
+  const keys = Object.keys(table.schema).map(key => key.toLowerCase())
+  return keys.indexOf(columnName.toLowerCase()) !== -1
+}
+
+export class RelationshipErrorChecker {
+  constructor(invalidThroughTableFn, relationshipExistsFn) {
+    this.invalidThroughTable = invalidThroughTableFn
+    this.relationshipExists = relationshipExistsFn
+  }
+
+  setType(type) {
+    this.type = type
+  }
+
+  isMany() {
+    return this.type === RelationshipTypes.MANY_TO_MANY
+  }
+
+  relationshipTypeSet(type) {
+    return !type ? noRelationshipType : null
+  }
+
+  tableSet(table) {
+    return !table ? tableNotSet : null
+  }
+
+  throughTableSet(table) {
+    return this.isMany() && !table ? tableNotSet : null
+  }
+
+  manyForeignKeySet(key) {
+    return this.isMany() && !key ? foreignKeyNotSet : null
+  }
+
+  foreignKeySet(key) {
+    return !this.isMany() && !key ? foreignKeyNotSet : null
+  }
+
+  primaryKeySet(key) {
+    return !this.isMany() && !key ? primaryKeyNotSet : null
+  }
+
+  throughIsNullable() {
+    return this.invalidThroughTable() ? throughNotNullable : null
+  }
+
+  doesRelationshipExists() {
+    return this.isMany() && this.relationshipExists()
+      ? relationshipAlreadyExists
+      : null
+  }
+
+  differentTables(table1, table2, table3) {
+    // currently don't support relationships back onto the table itself, needs to relate out
+    const error = table1 && (table1 === table2 || (table3 && table1 === table3))
+    return error ? mustBeDifferentTables : null
+  }
+
+  columnBeingUsed(table, column, ogName) {
+    return isColumnNameBeingUsed(table, column, ogName) ? columnCantExist : null
```

Comment:
NAB: Could change the `columnCantExist` variable to simply be `columnBeingUsed` ?